### PR TITLE
[PyGrain Migration] Add pure Python paths to audio converters for Grain pipelines

### PR DESCRIPTION
**Pull request by laxmareddyp**
Port the Whisper, Moonshine, Gemma3n and Gemma4 audio converters to NumPy so they can run inside Grain worker processes without TensorFlow, following the ImageConverter precedent.

- Add NumPy helpers mirroring tf.signal.frame / tf.signal.hann_window.
- Split each converter into _call_python (NumPy) and _call_tf, with call() dispatching on _allow_python_workflow / in_tf_function().
- Default _allow_python_workflow to True for all four converters.
- Store constants (mel filterbanks, windows, framing indices) as NumPy instead of backend or TF tensors so layers pickle into Grain workers.
- Return NumPy inside Grain via convert_preprocessing_outputs_python.
- Add Grain and TF/Python numeric parity tests.

## Checklist

- [x] I have read and followed the [PR contribution policy](https://github.com/keras-team/keras/issues/23601).
- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/gemma3n/gemma3n_audio_converter.py [hunk 1]
```diff
@@ -9,6 +9,13 @@
 from keras_hub.src.api_export import keras_hub_export
 from keras_hub.src.layers.preprocessing.audio_converter import AudioConverter
 from keras_hub.src.models.gemma3n.gemma3n_backbone import Gemma3nBackbone
+from keras_hub.src.utils.audio_utils import frame_signal
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/gemma3n/gemma3n_audio_converter.py:14]: Done, Thanks

File: keras_hub/src/models/gemma3n/gemma3n_audio_converter.py [hunk 2]
```diff
@@ -168,9 +178,9 @@ def __init__(
         if self.fft_overdrive:
             fft_length *= 2
         self.fft_length = fft_length
-        hann_arange = tf.range(self.frame_length, dtype=self.compute_dtype)
+        hann_arange = np.arange(self.frame_length, dtype=self.compute_dtype)
         self.window = 0.5 * (
-            1 - tf.cos(2 * np.pi * hann_arange / self.frame_length)
+            1 - np.cos(2 * np.pi * hann_arange / self.frame_length)
         )
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/gemma3n/gemma3n_audio_converter.py]: Done, Thanks

File: keras_hub/src/models/gemma4/gemma4_audio_converter.py [hunk 3]
```diff
@@ -4,6 +4,12 @@
 from keras_hub.src.api_export import keras_hub_export
 from keras_hub.src.layers.preprocessing.audio_converter import AudioConverter
 from keras_hub.src.models.gemma4.gemma4_backbone import Gemma4Backbone
+from keras_hub.src.utils.tensor_utils import (
+    convert_preprocessing_outputs_python,
+)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/gemma4/gemma4_audio_converter.py:10]: Done, Thanks

File: keras_hub/src/models/gemma4/gemma4_audio_converter.py [hunk 4]
```diff
@@ -121,17 +130,18 @@ def __init__(
         # HTK mel filterbank: shape (num_fft_bins // 2 + 1, num_mels).
         self.mel_filters = self._get_mel_filters()
 
-        # Periodic Hann window matching HF
+        # Periodic Hann window matching HF. Kept as NumPy so the layer holds
+        # no backend tensors (Grain pickles layers into worker processes).
         length = self.frame_length + 1
         window = np.hanning(length)
-        self.window = ops.convert_to_tensor(window[:-1], dtype="float32")
+        self.window = window[:-1].astype("float32")
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/gemma4/gemma4_audio_converter.py]: Done, Thanks

File: keras_hub/src/models/gemma3n/gemma3n_audio_converter.py [hunk 5]
```diff
@@ -207,21 +217,20 @@ def _create_filterbank_matrix(
...
+        return fb.astype(self.compute_dtype)
 
     def _extract_spectrogram(self, waveform, attention_mask):
-        waveform = tf.cast(waveform, dtype=self.compute_dtype)
+        waveform = np.asarray(waveform, dtype=self.compute_dtype)
         if self.dither > 0.0:
-            waveform = waveform + self.dither * tf.random.normal(
-                tf.shape(waveform), dtype=waveform.dtype
-            )
+            waveform = waveform + self.dither * np.random.default_rng(
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma3n/gemma3n_audio_converter.py:251]: randomness in preprocessing should be deterministic per record so results do not depend on worker count or resume point. An unseeded generator gives every worker its own stream and makes runs unreproducible. either defaults to 0.0 so this is rarely hit, but when it is set, derive the generator from a layer seed plus a stable hash of the waveform.

File: keras_hub/src/models/whisper/whisper_audio_converter.py [hunk 6]
```diff
@@ -208,7 +223,77 @@ def tf_log10(x):
...
+        """NumPy equivalent of `_extract_audio_features`."""
+        dtype = self.compute_dtype
+        # Use "reflection" padding - `tf.signal.stft` uses symmetric padding
+        # internally.
+        pad_width = self.num_fft_bins // 2
+        audio = np.pad(audio, ((0, 0), (pad_width, pad_width)), mode="reflect")
+
+        # Compute the mel spectrogram.
+        frames = frame_signal(audio, self.num_fft_bins, self.stride)
+        frames = frames * hann_window(self.num_fft_bins, dtype="float64")
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/whisper/whisper_audio_converter.py]: Multiplying float32 frames by a float64 window promotes the whole STFT and mel matmul to float64, doubling memory and compute. Use self.compute_dtype for the window.

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks! few nit comments

File: keras_hub/src/models/whisper/whisper_audio_converter.py (hunk 6 above)
Comment (laxmareddyp) [reply keras_hub/src/models/whisper/whisper_audio_converter.py]: Done — using `self.compute_dtype` now. Confirmed the promotion is gone: the
STFT drops from complex128 to complex64, halving the magnitudes buffer.

File: keras_hub/src/models/gemma3n/gemma3n_audio_converter.py (hunk 5 above)
Comment (laxmareddyp) [reply keras_hub/src/models/gemma3n/gemma3n_audio_converter.py:251]: Done — the generator is now derived from a layer `seed` plus a SHA-256 of the
record's unpadded samples, so dither is per-record and independent of batching.
Covered by `test_dither_determinism` on both the Python and TF paths.

**Review by divyashreepathihalli — APPROVED** [review]
